src/components/SearchView.tsx

```diff
@@ -29,23 +35,49 @@ export class SearchView extends React.Component<SearchViewProps, SearchViewState
     constructor(props: SearchViewProps) {
         super(props);
 
-        // set default state
-        this.state = ({
-            searchText: "",
-            displayMode: "list"
-        })
-
+        this.state = {
+            searchText: '',
+            displayMode: "list",
+            selectedCategories: this.props.categories,
+            structured: false,
+            detailed: false
+        };
         this.searchResultListItems = null;
     }
 
     getSearchText(): string {
         return this.state.searchText;
     }
 
+    onStructuredModeChanged(value: boolean) {
+        this.setState({ structured: value });
+    }
+
+    onDetailedModeChanged(value: boolean) {
+        this.setState({ detailed: value });
+    }
+
+    onCategoriesChanged(categories: string[]) {
+        this.setState({ selectedCategories: categories })
+    }
+
     generateStructuredItems(): JSX.Element[] {
+        let structuredItems: JSX.Element[] = [];
+        let categoryItems: ItemData[] = [];
+        
         let index = 0;
-        return this.props.sections.map((item: ItemData) =>
-            <LibraryItem key={index++} libraryContainer={this.props.libraryContainer} data={item} />);
+
+        this.props.sections.forEach(section =>
+            categoryItems = categoryItems.concat(section.childItems)
```

Comment:
Can we not use `categoryItems.push(...)` instead of `categoryItems.concat(...)`? The latter creates a new array altogether, we should use the fastest way.